File: LeetCode/set_matrix_zeroes.cpp
class Solution {
public:
    void solve(vector<vector<int>>& matrix, int i, int j) {
        int k;
        
        k = i;
        while(k >= 0)
            matrix[k--][j] = 0;
        
        k = i;
        while(k < matrix.size())
            matrix[k++][j] = 0;
        
        k = j;
        while(k >= 0)
            matrix[i][k--] = 0;
        
        k = j;
        while(k < matrix[i].size())
            matrix[i][k++] = 0;
    }
    
    void setZeroes(vector<vector<int>>& matrix) {
        vector< pair<int, int> > zeroesPos;
        
        for(int i = 0; i < matrix.size(); i++) {
            for(int j = 0; j < matrix[i].size(); j++) {
                if(!matrix[i][j])
                    zeroesPos.push_back(make_pair(i, j));
            }
        }
        
        for(int i = 0; i < zeroesPos.size(); i++) {
            solve(matrix, zeroesPos[i].first, zeroesPos[i].second);
        }
    }
};

File: LeetCode/set_matrix_zeroes.go
func setZeroes(matrix [][]int)  {
    n, m := len(matrix), len(matrix[0])
    rowZero, colZero := false, false

    for i := 0; i < n; i++ {
        for j := 0; j < m; j++ {
            if matrix[i][j] == 0 {
                if i == 0 {
                    rowZero = true
                } else {
                    matrix[i][0] = 0
                }

                if j == 0 {
                    colZero = true
                } else {
                    matrix[0][j] = 0
                }
            }
        }
    }

    for i := 1; i < n; i++ {
        for j := 1; j < m; j++ {
            if matrix[0][j] == 0 || matrix[i][0] == 0 {
                matrix[i][j] = 0
            }
        }
    }

    if rowZero {
        for j := 0; j < m; j++ {
            matrix[0][j] = 0
        }
    }

    if colZero {
        for i := 0; i < n; i++ {
            matrix[i][0] = 0
        }
    }
}
